### ISSUE TITLE
Conflict with the elastic-apm gem

### ISSUE THREAD

**Issue by paul-mesnilgrente** [airbrake/airbrake#1080]
## Describe the bug

I upgraded to the latest version of Airbrake (10.0.2) and Elastic-APM (3.6.0) and I started to have a `SystemStackError` in my controller. After some time I found that the stack was too deep. Logging the backtrace showed:

```
"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm/spies/net_http.rb:72:in `block in request'"
"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm.rb:273:in `with_span'"
"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm/spies/net_http.rb:62:in `request'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rails/net_http.rb:11:in `block in request'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rack.rb:26:in `block in capture_timing'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:29:in `span'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:18:in `block in span'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:18:in `tap'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:18:in `span'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rack.rb:25:in `capture_timing'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rails/net_http.rb:10:in `request'"

"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm/spies/net_http.rb:72:in `block in request'"
"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm.rb:273:in `with_span'"
"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm/spies/net_http.rb:62:in `request'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rails/net_http.rb:11:in `block in request'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rack.rb:26:in `block in capture_timing'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:29:in `span'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:18:in `block in span'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:18:in `tap'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-ruby-4.13.3/lib/airbrake-ruby/timed_trace.rb:18:in `span'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rack.rb:25:in `capture_timing'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rails/net_http.rb:10:in `request'"

"/app/vendor/cache/ruby/2.6.0/gems/elastic-apm-3.6.0/lib/elastic_apm/spies/net_http.rb:72:in `block in request'"
```

You can see here that elastic-apm is called by airbrake which is calling airbrake, and that in a loop

I tried the same code with elastic-apm 3.1.0 and airbrake 9.5.5 and it worked fine (that's the only other config I tried)

## Steps to reproduce
1. Install elastic-apm 3.6.0 (I only tested that one)
2. Install airbrake 10.0.2 (I only tested that one)
3. Run a code calling an API with RestClient. My last lines called before entering the loop:

```
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rack.rb:25:in `capture_timing'"
"/app/vendor/cache/ruby/2.6.0/gems/airbrake-10.0.2/lib/airbrake/rails/net_http.rb:10:in `request'"
"/app/vendor/cache/ruby/2.6.0/gems/rest-client-2.1.0/lib/restclient/request.rb:473:in `net_http_do_request'"
"/app/vendor/cache/ruby/2.6.0/gems/rest-client-2.1.0/lib/restclient/request.rb:733:in `block in transmit'"
"/usr/local/lib/ruby/2.6.0/net/http.rb:920:in `start'"
"/app/vendor/cache/ruby/2.6.0/gems/rest-client-2.1.0/lib/restclient/request.rb:727:in `transmit'"
"/app/vendor/cache/ruby/2.6.0/gems/rest-client-2.1.0/lib/restclient/request.rb:163:in `execute'"
"/app/vendor/cache/ruby/2.6.0/gems/rest-client-2.1.0/lib/restclient/request.rb:63:in `execute'"
```

## Expected behavior

No conflicts with elastic-apm.

## Environment

- **OS**: [ruby:2.6.5-alpine](https://hub.docker.com/_/ruby)
- **Ruby version**: 2.6.5
- **Framework and version**: Rails 5.1.7
- **APM Server version**: 
- **Agent version**: 3.6.0 (my gem version)

Related issue: https://github.com/elastic/apm-agent-ruby/issues/753

**Comment by kyrylo** [comment]
Hi, thanks for the report. `apm-agent-ruby` needs to stop using `class_eval`: https://github.com/elastic/apm-agent-ruby/blob/8a44d02d247887dd0c806282a68174c541783b76/lib/elastic_apm/spies/net_http.rb#L34

We had a similar problem where Airbrake was doing `class_eval` to monkey-patch `Net::HTTP`. This resulted in a conflicted with DataDog. Their gem was doing the right thing (using `prepend`), so I did the same with Airbrake. More info: https://github.com/airbrake/airbrake/issues/1077

**Comment by paul-mesnilgrente** [comment]
Hi, thanks for your quick response. I am not sure that this is solving my problem. The fix was deployed in 10.0.2 but this is the version that I was using...

**Comment by kyrylo** [comment]
`apm-agent-ruby` is the one who needs a fix (they should stop using `class_eval`) for monkey-patching.
